### PR TITLE
use timestamps in localization tools

### PR DESCRIPTION
**Pull request by sammlapp**
modifies the requirements for detections dataframe passed to SynchronizedRecorderArray.localize_detections(): multi-index should have fourth column "start_timestamp" with datetime.datetime. If given regular index (file, start_time, end_time), will try to parse audio file metadata to get timestamps.

This makes the API a bit harder to use for cases where the start_timestamp cannot be parsed automatically. However, it will work smoothly for audiomoth files and we can implement parsing of other metadata types as needed.

Also changes SpatialEvent's attribute and ._estimate_delays: attribute `.receiver_start_time_offsets` lists the start_time for each reciever in the event. Will all be equal only if if all audio files started at the same time. Alternatively, if `.receiver_start_time_offsets` is None, attempts to rely on `.start_timestamp` and use audio file's metadata to determine correct offset (via Audio.from_file(...start_timestamp=...)

### CODE REVIEW

File: opensoundscape/localization.py [hunk 1]
```diff
@@ -281,6 +337,106 @@ def _localize_after_cross_correlation(self, localization_algorithm):
...
+                f"Failed to retrieve recording start time from metadata of file {file}."
+            ) from exc
+        # return self.tz.localize(recording_start_dt + datetime.timedelta(seconds=start_time))
+
+        # if not (
+        #     type(recording_start_dt) == datetime.datetime
+        #     and recording_start_dt.tzinfo is not None
+        # ):
+        #     pass
+
```
Comment (louisfh) [inline opensoundscape/localization.py:435]: Please remove, or add a line explaining why this commented code is here.

**Review by louisfh — CHANGES_REQUESTED** [review]
This looks good. You already mention some in #TODO comments but I think we need tests for:
- Parsing timestamps automatically from audio files.
- Files with different receiver_start_time_offsets - if we are explicitly supporting this.
- SpatialEvent.from_dict, events_to_df and df_to_events methods

Do we need to think about if timestamps are timezone localized (e.g. UTC) or not? The only issues I can foresee in the code already are if some of the files have timezone localized timestamps, and others don't. `det_cls.groupby(level="start_timestamp"):` in group_events may not work as expected in this case, failing silently. It seems like a very unlikely edge case, and would be the users fault if they provide inconsistent timestamps in the detections dataframe. However given we enable automatic parsing of audio metadata we should be sure this isn't likely to happen with e.g. different firmware versions/recorder hardware. If we think its a real risk, we could add a check of tz localized when parsing.

**Comment by sammlapp** [comment]
> This looks good. You already mention some in #TODO comments but I think we need tests for:
> 
> * Parsing timestamps automatically from audio files.
> * Files with different receiver_start_time_offsets - if we are explicitly supporting this.
> * SpatialEvent.from_dict, events_to_df and df_to_events methods

Thanks, I forgot I didn't finish tests yet

> Do we need to think about if timestamps are timezone localized (e.g. UTC) or not? The only issues I can foresee in the code already are if some of the files have timezone localized timestamps, and others don't. `det_cls.groupby(level="start_timestamp"):` in group_events may not work as expected in this case, possibly failing silently. It seems like a very unlikely edge case, and would be the users fault if they provide inconsistent timestamps in the detections dataframe. However given we enable automatic parsing of audio metadata we should be sure this isn't likely to happen with e.g. different firmware versions/recorder hardware. If we think its a real risk, we could add a check of tz localized when parsing.

Good point - I think it might raise an error as soon as you try to compare localized and unlocalized timestamps, but I can check

**Comment by sammlapp** [comment]
I added the missing tests, and decided on a strict behavior for timezone awareness of datetime: it now requires that all datetimes are localized to a timezone, which could avoid some unforeseen silent error where the user doesn't think about the timezones of their timestamps

**Pull request merged**